File: 665-Non-decreasing-Array.py
class Solution(object):
    def checkPossibility(self, nums):
        
        ModifyingTimes = 0
        for i in range(1 , len(nums)) :

            if nums[i-1] > nums[i] :
                ModifyingTimes += 1

                # delete left one
                if nums[i-2] < nums[i] or i < 2 :
                    nums[i-1] = nums[i]
                # delete right one
                else:
                    nums[i] = nums[i-1]
                
            if ModifyingTimes > 1 :
                return False
        
        return True

File: LeoHuang/665-Non-decreasing-Array/665-Non-decreasing-Array.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var checkPossibility = function(nums) {
  let adjusted = false;
  for(let i = 0; i < nums.length; i++) {
    if(nums[i] > nums[i + 1]) {
      if(adjusted) {
        return false;
      }
      if(nums[i + 2] && nums[i + 1] <= nums[i + 2]) {
        if(nums[i - 1] && (nums[i - 1] > nums[i + 2] || nums[i - 1] > nums[i + 1]) && nums[i] > nums[i + 2]) {
          return false;
        }
      } else if(!nums[i + 2]) {
        
      } else {
        // nums[i + 1] > nums[i + 2]
        return false;
      }
      adjusted = true;
    }
  }
  return true;
};
